File: backend/app/auth/password_rules.py
"""Reglas de fortaleza para contraseñas nuevas.

Fuente de verdad única para la validación de contraseñas en el backend.
Reflejada en:
- web/src/assets/js/password-rules.js (formulario estático en 11ty)
- frontend/src/app/validators/password-strength.validator.ts (SPA Angular)

Reglas:
- Entre 8 y 20 caracteres (el máximo es secreto para el usuario).
- Al menos una letra mayúscula, una minúscula, un número
  y un carácter especial de la lista permitida.
"""

import re

MIN_LENGTH = 8
MAX_LENGTH = 20

# Lista de especiales sincronizada con password-rules.js y el validador Angular.
SPECIAL_CHARS = "!@#$%^&*-_=+.,;:?"

# Regex equivalente para uso externo (auditoría, tests cruzados).
PASSWORD_REGEX = re.compile(
    r"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*\-_=+.,;:?]).{8,20}$"
)

_RE_UPPER   = re.compile(r"[A-Z]")
_RE_LOWER   = re.compile(r"[a-z]")
_RE_DIGIT   = re.compile(r"\d")
_RE_SPECIAL = re.compile(r"[!@#$%^&*\-_=+.,;:?]")


def validate_password(pwd: str) -> str | None:
    """Valida una contraseña nueva contra las reglas de fortaleza.

    Parameters
    ----------
    pwd:
        Contraseña en claro propuesta por el usuario.

    Returns
    -------
    str | None
        ``None`` si la contraseña es válida, o un mensaje descriptivo
        del primer criterio incumplido.
    """
    if not isinstance(pwd, str) or not pwd:
        return "La contraseña es obligatoria."
    if len(pwd) < MIN_LENGTH:
        return f"La contraseña debe tener al menos {MIN_LENGTH} caracteres."
    if len(pwd) > MAX_LENGTH:
        return f"La contraseña no puede superar los {MAX_LENGTH} caracteres."
    if not _RE_UPPER.search(pwd):
        return "La contraseña debe contener al menos una letra mayúscula."
    if not _RE_LOWER.search(pwd):
        return "La contraseña debe contener al menos una letra minúscula."
    if not _RE_DIGIT.search(pwd):
        return "La contraseña debe contener al menos un número."
    if not _RE_SPECIAL.search(pwd):
        return "La contraseña debe contener al menos un carácter especial."
    return None

File: web/src/assets/js/password-rules.js
/**
 * password-rules.js
 *
 * Fuente de verdad única para la validación de fortaleza de contraseñas
 * en el sitio estático 11ty. Reflejada en:
 *   - backend/app/auth/password_rules.py
 *   - frontend/src/app/validators/password-strength.validator.ts
 *
 * El máximo de 20 caracteres es SECRETO: el criterio `maxLength` no debe
 * mostrarse en la lista de requisitos hasta que el usuario lo supere.
 */
'use strict';

(function (global) {
  var MIN_LENGTH = 8;
  var MAX_LENGTH = 20;
  var SPECIAL_CHARS = '!@#$%^&*-_=+.,;:?';
  var SPECIAL_RE = /[!@#$%^&*\-_=+.,;:?]/;

  var CRITERIA = [
    { id: 'minLength', label: 'Mínimo ' + MIN_LENGTH + ' caracteres',          test: function (v) { return v.length >= MIN_LENGTH; } },
    { id: 'upper',     label: 'Al menos 1 letra mayúscula',                    test: function (v) { return /[A-Z]/.test(v); } },
    { id: 'lower',     label: 'Al menos 1 letra minúscula',                    test: function (v) { return /[a-z]/.test(v); } },
    { id: 'digit',     label: 'Al menos 1 número',                             test: function (v) { return /\d/.test(v); } },
    { id: 'special',   label: 'Al menos 1 carácter especial (' + SPECIAL_CHARS + ')', test: function (v) { return SPECIAL_RE.test(v); } },
  ];

  // Criterio "maxLength" — oculto hasta que el usuario lo supere.
  var MAX_CRITERION = {
    id: 'maxLength',
    label: 'No superar ' + MAX_LENGTH + ' caracteres',
    test: function (v) { return v.length <= MAX_LENGTH; },
  };

  function validate(pwd) {
    var v = String(pwd == null ? '' : pwd);
    if (!v)               return 'La contraseña es obligatoria.';
    if (v.length < MIN_LENGTH) return 'La contraseña debe tener al menos ' + MIN_LENGTH + ' caracteres.';
    if (v.length > MAX_LENGTH) return 'La contraseña no puede superar los ' + MAX_LENGTH + ' caracteres.';
    if (!/[A-Z]/.test(v)) return 'La contraseña debe contener al menos una letra mayúscula.';
    if (!/[a-z]/.test(v)) return 'La contraseña debe contener al menos una letra minúscula.';
    if (!/\d/.test(v))    return 'La contraseña debe contener al menos un número.';
    if (!SPECIAL_RE.test(v)) return 'La contraseña debe contener al menos un carácter especial.';
    return null;
  }

  global.SS_PASSWORD_RULES = {
    MIN_LENGTH: MIN_LENGTH,
    MAX_LENGTH: MAX_LENGTH,
    CRITERIA: CRITERIA,
    MAX_CRITERION: MAX_CRITERION,
    validate: validate,
  };
}(typeof window !== 'undefined' ? window : globalThis));
